langkit/templates/iterator_types_ada.mako

```diff
@@ -0,0 +1,187 @@
+## vim: filetype=makoada
+
+<%def name="public_api_decl(cls)">
+  type ${cls.api_name} is tagged private;
+
+  function Next
+    (I : ${cls.api_name};
+     E : out ${cls.element_type.api_name}) return Boolean;
+</%def>
+
+<%def name="public_api_private_decl(cls)">
+  type ${cls.api_name} is new Ada.Finalization.Controlled with record
+     Internal_Iterator : Implementation.${cls.name};
+  end record;
+
+  overriding procedure Initialize (T : in out ${cls.api_name});
+  overriding procedure Adjust     (T : in out ${cls.api_name});
+  overriding procedure Finalize   (T : in out ${cls.api_name});
+</%def>
+
+<%def name="incomplete_decl(cls)">
+   type ${cls.iterator_type_name};
+   type ${cls.name} is access all ${cls.iterator_type_name};
+</%def>
+
+<%def name="ada_api_converters_decl(cls)">
+   % if cls.to_public_converter_required:
+      function ${cls.to_public_converter}
+         (Value : ${cls.name}) return ${cls.api_name};
+   % endif
+</%def>
+
+<%def name="public_api_body(cls)">
+   % if cls.to_public_converter_required:
+      function ${cls.to_public_converter}
+         (Value : ${cls.name}) return ${cls.api_name} is
+      begin
+         Inc_Ref (Value);
+         return (Ada.Finalization.Controlled with Internal_Iterator => Value);
+      end;
+   % endif
+
+  ----------
+  -- Next --
+  ----------
+
+  function Next
+    (I : ${cls.api_name};
+     E : out ${cls.element_type.api_name}) return Boolean
+  is
+     Internal_Element : ${cls.element_type.name};
+     Result           : Boolean :=
+        Next (I.Internal_Iterator, Internal_Element);
+  begin
+     if Result then
+        E := ${cls.element_type.to_public_expr('Internal_Element')};
+     end if;
+     return Result;
+  end Next;
+
+  ----------------
+  -- Initialize --
+  ----------------
+
+  procedure Initialize (T : in out ${cls.api_name}) is
+  begin
+     T.Internal_Iterator := null;
+  end Initialize;
+
+  ------------
+  -- Adjust --
+  ------------
+
+  overriding procedure Adjust (T : in out ${cls.api_name}) is
+  begin
+     Inc_Ref (T.Internal_Iterator);
+  end Adjust;
+
+  --------------
+  -- Finalize --
+  --------------
+
+  procedure Finalize   (T : in out ${cls.api_name}) is
+  begin
+     Dec_Ref (T.Internal_Iterator);
+     T.Internal_Iterator := null;
+  end Finalize;
+</%def>
+
+<%def name="decl(cls)">
+
+   <% elt_type = cls.element_type.name %>
+
+   type ${cls.iterator_type_name} is record
+      Ref_Count : Integer;
+      --  Negative values are interpreted as "always living singleton".
+      --  Non-negative values have the usual ref-counting semantics.
```

Comment:
IIUC we never need always living singletons for iterators. If so, can you update the comment accordingly?